android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageFacebookConceal.java

```diff
@@ -0,0 +1,96 @@
+package com.oblador.keychain.cipherStorage;
+
+import android.os.Build;
+
+import com.facebook.android.crypto.keychain.AndroidConceal;
+import com.facebook.android.crypto.keychain.SharedPrefsBackedKeyChain;
+import com.facebook.crypto.Crypto;
+import com.facebook.crypto.CryptoConfig;
+import com.facebook.crypto.Entity;
+import com.facebook.crypto.keychain.KeyChain;
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.oblador.keychain.exceptions.CryptoFailedException;
+
+import java.nio.charset.Charset;
+
+public class CipherStorageFacebookConceal implements CipherStorage {
+    public static final String CIPHER_STORAGE_NAME = "FacebookConceal";
+    public static final String KEYCHAIN_DATA = "RN_KEYCHAIN";
+    public static final String EMPTY_STRING = "";
+    private final Crypto crypto;
+
+    public CipherStorageFacebookConceal(ReactApplicationContext reactContext) {
+        KeyChain keyChain = new SharedPrefsBackedKeyChain(reactContext, CryptoConfig.KEY_256);
+        this.crypto = AndroidConceal.get().createDefaultCrypto(keyChain);
+    }
+
+    @Override
+    public String getCipherStorageName() {
+        return CIPHER_STORAGE_NAME;
+    }
+
+    @Override
+    public int getMinSupportedApiLevel() {
+        return Build.VERSION_CODES.JELLY_BEAN;
+    }
+
+    @Override
+    public EncryptionResult encrypt(String service, String username, String password) throws CryptoFailedException {
+        if (!crypto.isAvailable()) {
+            throw new CryptoFailedException("Crypto is missing");
+        }
+        service = service == null ? EMPTY_STRING : service;
+
+        Entity usernameEntity = createUsernameEntity(service);
+        Entity passwordEntity = createPasswordEntity(service);
+
+        try {
+            byte[] encryptedUsername = crypto.encrypt(username.getBytes(Charset.forName("UTF-8")), usernameEntity);
+            byte[] encryptedPassword = crypto.encrypt(password.getBytes(Charset.forName("UTF-8")), passwordEntity);
+
+            return new EncryptionResult(encryptedUsername, encryptedPassword, this);
+        } catch (Exception e) {
+            throw new CryptoFailedException("Encryption failed for service " + service, e);
+        }
+    }
+
+    @Override
+    public DecryptionResult decrypt(String service, byte[] username, byte[] password) throws CryptoFailedException {
+        if (!crypto.isAvailable()) {
+            throw new CryptoFailedException("Crypto is missing");
+        }
+        service = service == null ? EMPTY_STRING : service;
```

Comment:
there are three places where `service` enters the module and from where it is passed onward - the methods for `get`, `set` and `reset` password. These are unlikely to change. I suggest that the `service = service == null ? EMPTY_STRING : service;` is done there so that we're really sure that we don't have to do a null check in any other methods of the module - for example in `getEntityPrefix`. It also means less duplication, since now all implementations of `CipherStorage` have to do this check.